### PR TITLE
Add synchronization around SDLResponseDispatcher

### LINKED ISSUE #1515: rpcResponseHandlerMap SIGABRT crash in the sdl library
I got a crash in the sdl library.
The screen is from Firebase & Cocoapod.
I think the root cause is the variable rpcResponseHandlerMap is nil.
Could you check it again?

image

image

##### Reproduction Steps
1. Launch the app
2. Connect with the sdl device
3. Streaming the video

##### OS & Version Information
* iOS Version: 13.3.0
* Mobile: iPhone 11
* SDL iOS Version: 6.4.1

### PR DESCRIPTION
**Pull request by joeljfischer**
Fixes #1515 

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

#### Unit Tests
No unit tests were added, they were run and pass.

#### Core Tests
* Tested example app with numerous menu cells, submenu cells, soft buttons, etc. to stress test RPCs waiting for synchronization.

Core version / branch / commit hash / module tested against: Sync 3.0 (19205_DEVTEST), Manticore v2.4.2 (SDL Core v6.0.1)
HMI name / version / branch / commit hash / module tested against: Sync 3.0 (19205_DEVTEST), Manticore v2.4.2 (Generic HMI v0.7.2.)

### Summary
* Methods that alter the internal dictionaries and map tables of the `SDLResponseDispatcher` are now synchronized.

##### Bug Fixes
* Fix concurrent setting to dictionaries and map tables causing crashes by synchronizing setter access.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLink/SDLResponseDispatcher.m [hunk 1]
```diff
@@ -86,67 +86,72 @@ - (instancetype)initWithNotificationDispatcher:(nullable id)dispatcher {
...
+            SDLButtonName buttonName = subscribeButton.buttonName;
+            if (!buttonName) {
+                @throw [NSException sdl_missingIdException];
+            }
+            if (subscribeButton.handler) {
+                self.buttonHandlerMap[buttonName] = subscribeButton.handler;
+            }
+        } else if ([request isKindOfClass:[SDLAlert class]]) {
+            SDLAlert *alert = (SDLAlert *)request;
+            [self sdl_addToCustomButtonHandlerMap:alert.softButtons];
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLResponseDispatcher.m]: The `sdl_runHandlerForButton` method also accesses the `customButtonHandlerMap` but its access is not restricted.

File: SmartDeviceLink/SDLResponseDispatcher.m [hunk 2]
```diff
@@ -86,67 +86,72 @@ - (instancetype)initWithNotificationDispatcher:(nullable id)dispatcher {
...
+            }
+        } else if ([request isKindOfClass:[SDLSubscribeButton class]]) {
+            // Convert SDLButtonName to NSString, since it doesn't conform to <NSCopying>
+            SDLSubscribeButton *subscribeButton = (SDLSubscribeButton *)request;
+            SDLButtonName buttonName = subscribeButton.buttonName;
+            if (!buttonName) {
+                @throw [NSException sdl_missingIdException];
+            }
+            if (subscribeButton.handler) {
+                self.buttonHandlerMap[buttonName] = subscribeButton.handler;
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLResponseDispatcher.m]: The `sdl_runHandlerForButton` method also accesses the `buttonHandlerMap` but its access is not restricted.

File: SmartDeviceLink/SDLResponseDispatcher.m [hunk 3]
```diff
@@ -86,67 +86,72 @@ - (instancetype)initWithNotificationDispatcher:(nullable id)dispatcher {
...
+        NSNumber *correlationId = request.correlationID;
+
+        // Check for RPCs that require an extra handler
+        if ([request isKindOfClass:[SDLAddCommand class]]) {
+            SDLAddCommand *addCommand = (SDLAddCommand *)request;
+            if (!addCommand.cmdID) {
+                @throw [NSException sdl_missingIdException];
+            }
+            if (addCommand.handler) {
+                self.commandHandlerMap[addCommand.cmdID] = addCommand.handler;
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLResponseDispatcher.m]: The `sdl_runHandlerForCommand` method also accesses the `commandHandlerMap` but its access is not restricted.

File: SmartDeviceLink/SDLResponseDispatcher.m (hunk 1 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLResponseDispatcher.m]: I re-did the whole sync model

File: SmartDeviceLink/SDLResponseDispatcher.m (hunk 2 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLResponseDispatcher.m]: I re-did the whole sync model

File: SmartDeviceLink/SDLResponseDispatcher.m (hunk 3 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLResponseDispatcher.m]: I re-did the whole sync model

**Comment by NicoleYarroch** [comment]
I am getting anything random failing test cases (between 3 and 8 total) in `SDLResponseDispatcherSpec`.

**Comment by joeljfischer** [comment]
There were issues due to directly accessing properties for a value that was stored in `storeRequest:` which stores it asynchronously. I re-did the tests to be "eventually" instead.

**Pull request merged**